src/openzeppelin/account.cairo

```diff
@@ -0,0 +1,233 @@
+    ) -> felt252;
+    #[external]
+    fn set_public_key(new_public_key: felt252);
+    #[view]
+    fn get_public_key() -> felt252;
+    #[view]
+    fn is_valid_signature(message: felt252, signature: Array<felt252>) -> u32;
+    #[view]
+    fn supports_interface(interface_id: u32) -> bool;
+}
+
+#[account_contract]
+mod Account {
+    use array::SpanTrait;
+    use array::ArrayTrait;
+    use box::BoxTrait;
+    use ecdsa::check_ecdsa_signature;
+    use serde::ArraySerde;
+    use starknet::get_tx_info;
+    use starknet::get_caller_address;
+    use starknet::get_contract_address;
+    use option::OptionTrait;
+    use zeroable::Zeroable;
+
+    use super::Call;
+    use super::ERC165_ACCOUNT_ID;
+    use super::ERC1271_VALIDATED;
+    use super::IAccount;
+    use super::TRANSACTION_VERSION;
+    use super::QUERY_VERSION;
+
+    use openzeppelin::introspection::erc165::ERC165;
+    use openzeppelin::utils::check_gas;
+    use openzeppelin::utils::span_to_array;
+
+    impl Account of IAccount {
+        fn __execute__(mut calls: Array<Call>) -> Array<Array<felt252>> {
+            // avoid calls from other contracts
+            // https://github.com/OpenZeppelin/cairo-contracts/issues/344
+            let sender = get_caller_address();
+            assert(sender.is_zero(), 'Account: invalid caller');
+
+            // check tx version
+            let tx_info = get_tx_info().unbox();
+            let version = tx_info.version;
+            if version != TRANSACTION_VERSION { // > operator not defined for felt252
+                assert(version == QUERY_VERSION, 'Account: invalid tx version');
+            }
+
+            _execute_calls(calls)
+        }
+
+        fn __validate__(mut calls: Array<Call>) -> felt252 {
+            _validate_transaction()
+        }
+
+        fn __validate_declare__(class_hash: felt252) -> felt252 {
+            _validate_transaction()
+        }
+
+        fn __validate_deploy__(
+            class_hash: felt252, contract_address_salt: felt252, _public_key: felt252
+        ) -> felt252 {
+            _validate_transaction()
+        }
+
+        fn set_public_key(new_public_key: felt252) {
+            _assert_only_self();
+            public_key::write(new_public_key);
+        }
+
+        fn get_public_key() -> felt252 {
+            public_key::read()
+        }
+
+        fn is_valid_signature(message: felt252, signature: Array<felt252>) -> u32 {
+            if _is_valid_signature(message, signature.span()) {
+                ERC1271_VALIDATED
+            } else {
+                0_u32
+            }
+        }
+
+        fn supports_interface(interface_id: u32) -> bool {
+            ERC165::supports_interface(interface_id)
+        }
+    }
+
+    struct Storage {
+        public_key: felt252
+    }
+
+    #[constructor]
+    fn constructor(_public_key: felt252) {
+        ERC165::register_interface(ERC165_ACCOUNT_ID);
```

Comment:
It certainly looks a little weird to call the constructor manually, but I think this is the way from the fact that we don't have proper OO inheritance in Cairo. I can't think as of now of any improvement that an initializer could bring over this. @andrew-fleming Interested in your thoughts as well.

Reply:
Hmm it definitely looks a little weird. I went back and forth a few times on this. I think using an `initializer` hardens the design of our current contracts (`ownable`, `pausable`, etc.) that also use `initializer`s. I'm not sure I see a benefit outside of that. It is extra code and arguably redundant in this case.

So, I'm not opposed to the `initializer` approach, but I lean toward just using the `constructor`

Reply:
> because that makes more sense for reusable modules which are not meant to be deployed, as constructor indicates.

But the example you provided is not this case, right? Account is meant to be deployed, so it make sense to have a constructor, and it makes sense to me to call this constructor in MyCustomAccount. This is similar to what happens in Solidity with constructors and inheritance:

```
contract Derived2 is Base {
    constructor(uint y) Base(y) {}
}
```

But in this case, we don't have inheritance, so we call the constructor manually.

Initializers still make sense for "not meant to be deployed" modules, but those are two separated cases.

Reply:
Agreed it's a pickle...I don't love either way. Defining `initializer` and `constructor` in the same mod isn't great (also why I leaned toward just having a `constructor`), but there's a better design element to using `initializer`s for mods and `constructor`s for as-is contracts IMO.

> what about leaving constructors for presets only, but initialization logic in initialize

+1